### PR TITLE
Handle model migration/upgrade when using Fan networking method

### PR DESCRIPTION
**Pull request by nvinuesa**
Since we are removing the fan support on 4.0, and therefore the fanconfigurer worker and facades are being removed, we must ensure that 3.6->4.0 migrations continue working. By checking if the error is NotImplemented, we can avoid failing in the fanconfigurer worker and allow the migrations to continue.

Also to fix the fact that we need to do something with the models that currently use the fan networking, we have decided to prevent the migration/upgrade of models with the config container-networking-method=fan *only if* there are containers deployed. This is done by adding a new validation method on model migration/upgrade.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [ ] ~Go unit tests, with comments saying what you're testing~
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## Blocked model
The first scenario is to try migrating a model that uses the Fan networking *and* has a container deployed. This migration should be blocked.

```
# (sudo snap install juju_4 --channel 4.0/edge)
juju_4 bootstrap aws c40
juju bootstrap aws c36                     
juju add-model m
juju model-config container-networking-method
fan
juju deploy ubuntu --to lxd
# Once it finished deploying, migrate it to c40:
juju migrate m c40
ERROR source prechecks failed: cannot migrate to controller due to issues:
"admin/m":
- cannot migrate models with container-networking-method=fan and containers deployed
```
## Happy model
This scenario is almost the same, but instead of deploying `ubuntu` to a container, it's deployed to a machine and therefore the same migration isn't blocked:

```
# (sudo snap install juju_4 --channel 4.0/edge)
juju_4 bootstrap aws c40
juju bootstrap aws c36                     
juju add-model m
juju model-config container-networking-method
fan
juju deploy ubuntu 
# Once it finished deploying, migrate it to c40:
juju migrate m c40
Migration started with ID "53d9d006-ccd9-4b6b-8824-e7b084a5828e:0"
```
At this point, you should be able to check the correct status of the migrated model.

## Links

**Jira card:** JUJU-6061

### CODE REVIEW

File: worker/fanconfigurer/fanconfigurer.go [hunk 1]
```diff
@@ -44,7 +44,10 @@ func (fc *FanConfigurer) processNewConfig() error {
 	defer fc.mu.Unlock()
 
 	fanConfig, err := fc.config.Facade.FanConfig()
-	if err != nil {
+	// NOTE(nvinuesa - 2024/05/27): This worker and this facade is being
+	// removed in 4.0, and therefore we check for NotImplemented errors
+	// and not fail in that case, to keep 3.6->4.0 migrations working.
+	if err != nil && !errors.Is(err, errors.NotImplemented) {
```
Comment (SimonRichardson) [inline worker/fanconfigurer/fanconfigurer.go:50]: I think we should log out that the implementation is not being used.

Comment (nvinuesa) [reply worker/fanconfigurer/fanconfigurer.go]: Just added a trace, I'm not sure this is 100% useful though. The only case this appear is a migration to 4.0.

File: worker/fanconfigurer/fanconfigurer.go [hunk 2]
```diff
@@ -45,7 +45,14 @@ func (fc *FanConfigurer) processNewConfig() error {
 
 	fanConfig, err := fc.config.Facade.FanConfig()
 	if err != nil {
```
Comment (hpidcock) [inline worker/fanconfigurer/fanconfigurer.go:47]: Consider refactoring into one if-else statement.
```
fanConfig, err := fc.config.Facade.FanConfig()
if errors.Is(err, errors.NotImplemented) {
	// NOTE(nvinuesa - 2024/05/27): This worker and this facade is being
	// removed in 4.0, and therefore we check for NotImplemented errors
	// and not fail in that case, to keep 3.6->4.0 migrations working.
	logger.Tracef("The fanconfigurer facade is not implemented, this is probably a migration from 3.6 to 4.0 in which case this is normal")
} else if err != nil {
	return err
}
```

Comment (nvinuesa) [reply worker/fanconfigurer/fanconfigurer.go]: Sure, done.

**Comment by jack-w-shaw** [comment]
This breaks 3.6 -> 3.6 migrations

```
$ juju bootstrap aws/eu-west-2 aws-dst
$ juju bootstrap aws/eu-west-2 aws-src
$ juju add-model m
$ juju deploy ubuntu --to lxd
$ juju migrate m aws-dst
ERROR source prechecks failed: cannot migrate to controller due to issues:
"admin/m":
- cannot migrate models with container-networking-method=fan and containers deployed
```

File: upgrades/upgradevalidation/shim.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+package upgradevalidation
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/replicaset/v3"
+
+	"github.com/juju/juju/state"
+)
+
+type upgradevalidationShim struct {
```
Comment (manadart) [inline upgrades/upgradevalidation/shim.go:13]: This is unused. You can remove it.

**Review by manadart — COMMENTED** [review]
Changes needed to factor the target version.

**Pull request closed without merging**